server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,522 @@
+        self.last_published = distributor["last_publish"]
+        self.last_deleted = repo.last_unit_removed
+        self.repo_type = repo.notes["_repo-type"]
+        self.predistributor = self._get_predistributor()
+        if self.last_published and not isinstance(self.last_published, basestring):
+            string_date = dateutils.format_iso8601_datetime(self.last_published)
+        elif self.last_published:
+            string_date = self.last_published
+        else:
+            string_date = None
+        if self.predistributor:
+            search_params = {'repo_id': repo.id,
+                             'distributor_id': self.predistributor["id"],
+                             'started': {"$gte": string_date}}
+            self.predist_history = RepoPublishResult.get_collection().find(search_params)
+        else:
+            self.predist_history = []
+
+        self.remote_path = self.get_remote_repo_path()
+
+        if not self.is_fastforward():
+            date_filter = None
+        else:
+            range_criteria = {START_DATE_KEYWORD: self.last_published,
+                              END_DATE_KEYWORD: self.predistributor["last_publish"]}
+            date_filter = self.create_date_range_filter(range_criteria)
+
+        self.extra_path = self._get_extra_path()
+        self.original_lists = {}
+        self.link_lists = {}
+
+        self._add_necesary_steps(date_filter=date_filter, config=config)
+
+    def is_fastforward(self):
+        """
+        This method checks whether this publish should be a fastforward publish.
+
+        :return: Whether or not this publish should be in fast forward mode
+        :rtype: bool
+        """
+        skip_fast_forward = False
+        for entry in self.predist_history:
+            str_skip_ff = entry.get("distributor_config", {}).get("skip_fast_forward", False)
+            skip_fast_forward |= misc.strbool2bool(str_skip_ff)
+            if entry.get("result", "error") == "error":
+                skip_fast_forward = True
+
+        if self.last_published and isinstance(self.last_published, basestring):
+            last_published = dateutils.parse_iso8601_datetime(self.last_published)
+        else:
+            last_published = self.last_published
+        if last_published:
+            last_published = last_published.replace(tzinfo=None)
+
+        if self.last_deleted and isinstance(self.last_deleted, basestring):
+            last_deleted = dateutils.parse_iso8601_datetime(self.last_deleted)
+        else:
+            last_deleted = self.last_deleted
+        if last_deleted:
+            last_deleted = last_deleted.replace(tzinfo=None)
+
+        config_skip_ff = self.get_config().get("skip_fast_forward", False)
+        skip_fast_forward = skip_fast_forward | misc.strbool2bool(config_skip_ff)
+        force_refresh = self.get_config().get("force_refresh", False)
+
+        return last_published and ((last_deleted and last_published > last_deleted) or
```

Comment:
This would be easier to read if we could pull some of it out into an independent evaluation. For example:

```
removal_since_last_publish = (last_deleted and last_published > last_deleted) or not last_deleted
```

That would make it easier to reason about whether the logic is correct.